src/box/sql.c

```diff
@@ -1581,3 +1574,114 @@ sql_constraint_drop(uint32_t space_id, const char *name)
 	region_truncate(region, used);
 	return rc;
 }
+
+/**
+ * Create new constraint in space.
+ *
+ * @param name Name of the constraint.
+ * @param space_id ID of the space.
+ * @param path JSON-path of the new constraint.
+ * @param value Encoded value of the new constraint.
+ */
+static int
+sql_constraint_create(const char *name, uint32_t space_id, const char *path,
+		      const char *value)
+{
+	struct region *region = &fiber()->gc;
+	uint32_t used = region_used(region);
+	char key[16];
+	char *key_end = key + mp_format(key, 16, "[%u]", space_id);
+	/*
+	 * Even if there were no constraints of this type, it is possible that
+	 * _space contains a non-empty field of this type with an empty map as
+	 * its value, which affects the update operation.
+	 */
+	struct tuple *tuple;
+	if (box_index_get(BOX_SPACE_ID, 0, key, key_end, &tuple) != 0) {
+		region_truncate(region, used);
+		return -1;
+	}
+	assert(tuple != NULL);
+	uint32_t path_len = strlen(path);
+	uint32_t path_hash = field_name_hash(path, path_len);
+	const char *field = tuple_field_raw_by_full_path(
+		tuple_format(tuple), tuple_data(tuple), tuple_field_map(tuple),
+		path, path_len, path_hash, TUPLE_INDEX_BASE);
+	bool is_empty = field == NULL;
+
+	const char *ops;
+	size_t ops_size;
+	if (is_empty) {
+		ops = mp_format_on_region(region, &ops_size, "[[%s%s{%s%p}]]",
```

Comment:
If the formatting on region fails, `ops` is `NULL`. Then `box_update()` below will crash because it always expects the opts to be a valid msgpack array. I suggest to move the region truncation to the end and make `goto` there from all error places so as not to duplicate the truncation too many times.

Reply:
Thanks, fixed!